### PR TITLE
CORDA-3601 Record a flow's finish time

### PR DESCRIPTION
**Pull request by lankydan**
Record a flow's finish time by updating its metadata record. It is set
in `updateCheckpoint` by checking the status of the checkpoint. If it is
`COMPLETED` it will set the `finishInstant` on the metadata object and
 update it.

### CODE REVIEW

**Review by williamvigorr3 — DISMISSED** [review]
LGTM!

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 1]
```diff
@@ -355,6 +356,24 @@ class FlowFrameworkTests {
...
+    fun `Flow metadata finish time is set in database when the flow finishes`() {
+        val terminationSignal = Semaphore(0)
+        val flow = aliceNode.services.startFlow(NoOpFlow(terminateUponSignal = terminationSignal))
+        mockNet.waitQuiescent()
+        aliceNode.database.transaction {
+            val metadata = session.find(DBCheckpointStorage.DBFlowMetadata::class.java, flow.id.uuid.toString())
+            assertNull(metadata.finishInstant)
+        }
+        terminationSignal.release()
+        mockNet.waitQuiescent()
```
Comment (kyriathar) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:373]: Maybe we could get rid of this `waitQuiescent()` if we `getOrThrow` instead on the returned fiber's future.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 2]
```diff
@@ -380,13 +380,20 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
         //val result = updateDBFlowResult(entity, checkpoint, now)
         val exceptionDetails = updateDBFlowException(entity, checkpoint, now)
 
+        val metadata = entity.flowMetadata.apply {
+            if (checkpoint.status == FlowStatus.COMPLETED && finishInstant == null) {
+                finishInstant = now
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:389]: Interesting that we initialise `now` to `Instant.now()`.  `DBCheckpointStorage` should look at the `clock` on `ServiceHub`.  Can we pass that into the constructor and use `clock.instant()` (I think!) please?

Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:389]: I think we  should update for all "finish types", so COMPLETED/FAILED/KILLED

Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:389]: I have some changes locally to correct that (the clock stuff). Going to do it in a separate PR, was waiting for this to go in.

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**